sound/soc/intel/boards/sof_sdw.c

```diff
@@ -1371,6 +1393,55 @@ static void set_dailink_map(struct snd_soc_dai_link_codec_ch_map *sdw_codec_ch_m
 		sdw_codec_ch_maps[i].connected_cpu_id = i / step;
 }
 
+static inline int find_codec_info_dai(const char *dai_name, int *dai_index)
+{
+	int i, j;
+
+	for (i = 0; i < ARRAY_SIZE(codec_info_list); i++) {
+		for (j = 0; j < codec_info_list[i].dai_num; j++) {
+			if (!strcmp(codec_info_list[i].dais[j].dai_name, dai_name)) {
+				*dai_index = j;
+				return i;
+			}
+		}
+	}
+
+	return -EINVAL;
+
+}
+
+static int sof_sdw_rtd_init(struct snd_soc_pcm_runtime *rtd)
+{
+	struct sof_sdw_codec_info *codec_info;
+	struct snd_soc_dai *dai;
+	int codec_index;
+	int dai_index;
+	int ret;
+	int i;
+
+	for_each_rtd_codec_dais(rtd, i, dai) {
+		codec_index = find_codec_info_dai(dai->name, &dai_index);
+		if (codec_index < 0) {
+			return -EINVAL;
+		}
+		codec_info = &codec_info_list[codec_index];
+		/*
+		 * A codec dai can be connected to different dai links for capture and playback,
+		 * but we only need to call the rtd_init function once.
```

Comment:
this is super confusing. The rtd_init function uses the 'rtd' parameter, so could it be that the rtd_init() behaves differently depending on the order in which dais are declared.

This goes back to the other comment on the 'initialized' field, we should try to make this a bit cleaner IMHO.

Reply:
The rtd_init will be called based on the order of codec dais. However, the order does not really matter. Typically, we append card->components, add card controls, dapm controls, dapm routes in rtd_init functions.  It doesn't matter if we add A first or B first. But, I do have an open that two rtd_init may add controls with the same name. For example, rt712 and rt1316 both add `SOC_DAPM_PIN_SWITCH("Speaker")`. 
Regarding the 'initialized' field, maybe rename to `rtd_init_done`, and add a comment to describe it?

Reply:
oh, ok. that would work if the codec dais are indeed on the same dailink. Otherwise I don't see how this would work since the same dai would be used for different purposes, and then the rtd_init_done would make no sense.

Do I get this right now? If yes, it's worth adding a big fat comment on what this rtd_init_done covers and when it would not work.

Reply:
it goes back to the change we made: all instances of a dai need to be in the same rtd/dailink.

We don't support e.g. two headset codecs used independently, or two identical amplifiers handling different content on separate links.

Reply:
> it goes back to the change we made: all instances of a dai need to be in the same rtd/dailink.
> 
> We don't support e.g. two headset codecs used independently, or two identical amplifiers handling different content on separate links.

No, we don't. The only case a dai can be in different dailinks is that it is in a capture dailink and a playback dailink.
For example, `SDW0-Playback` and `SDW0-Capture` dailinks both uses `rt711-sdca-aif1`.

Let me explain the PR with some examples.

The existing code assumes all codecs in a dailink are the same model.
So, we set dai_link.init = xxx_rtd_init in the codec dai.init function. For example, 
Say we have rt711-sdca jack and rt1316 amp.
`SDW0-Playback` dailink .init = rt_sdca_jack_rtd_init
`SDW0-Capture` dailink .init = NULL    #because `sof_sdw_rt_sdca_jack_init` will only set dai_link.init = rt_sdca_jack_rtd_init when playback is true
`SDW1-Playback` dailink .init = all_spk_init
`SDW1-Capture` dailink .init = NULL #because `sof_sdw_rt_amp_init` will only set dai_link.init = sof_sdw_rt_amp_init when playback is true

The PR is trying to remove the restriction of using the same codec model in a dailink, so, we need to call the xxx_rtd_init functions for different codec models in a dailink.
With the change, the dailink.init are all `sof_sdw_rtd_init`, and `sof_sdw_rtd_init` will call xxx_rtd_init for each codec model in the dailink. Use the same example, we will have
`SDW0-Playback` dailink .init = sof_sdw_rtd_init which will call rt_sdca_jack_rtd_init and set `rt711-sdca-aif1` dai.rtd_init_done = true
`SDW0-Capture` dailink .init = sof_sdw_rtd_init which will call nothing since `rt711-sdca-aif1` dai.rtd_init_done is true.
`SDW1-Playback` dailink .init = sof_sdw_rtd_init which will call `rt_amp_spk_rtd_init` once since `rt1316-aif` dai.rtd_init_done will be true after the first `rt_amp_spk_rtd_init` is called.
`SDW1-Capture` dailink .init = sof_sdw_rtd_init which will call nothing since `rt1316-aif` dai.rtd_init_done is true.

If we have rt712-sdca jack on `SDW0-Playback` and `SDW0-Capture` dailinks and rt712-sdca amp + rt1316 amp on `SDW1-Playback` and `SDW1-Capture` dailinks, we will have
`SDW0-Playback` dailink .init = sof_sdw_rtd_init which will call rt_sdca_jack_rtd_init and set `rt712-sdca-aif1` dai.rtd_init_done = true
`SDW0-Capture` dailink .init = sof_sdw_rtd_init which will call nothing since `rt712-sdca-aif1` dai.rtd_init_done is true.
`SDW1-Playback` dailink .init = sof_sdw_rtd_init which will call `rt_amp_spk_rtd_init` once since and 
`rt712_spk_rtd_init`
`SDW1-Capture` dailink .init = sof_sdw_rtd_init which will call nothing since `rt712-sdca-aif1` dai.rtd_init_done and `rt1316-aif` dai.rtd_init_done are both true.

Reply:
Are we confident that we only need to call rtd_init once? I could well imagine a situation, where the same device could be used in two different DAI links, thus requiring different slightly different init.

> it goes back to the change we made: all instances of a dai need to be in the same rtd/dailink.
> 
> We don't support e.g. two headset codecs used independently, or two identical amplifiers handling different content on separate links.

Do you have any more info on what change this was? I am curious as to what is introducing these limitations.

Reply:
> Are we confident that we only need to call rtd_init once? I could well imagine a situation, where the same device could be used in two different DAI links, thus requiring different slightly different init.

Yes, we use predefined be_id, all amp dai link will use the same BE id. IOW, we can't support amps on different dai links on the existing code. That is a restriction of the machine driver.

> 
> > it goes back to the change we made: all instances of a dai need to be in the same rtd/dailink.
> > We don't support e.g. two headset codecs used independently, or two identical amplifiers handling different content on separate links.
> 
> Do you have any more info on what change this was? I am curious as to what is introducing these limitations.

The change was to add the capability to support different codec models. For example, say if Cohen supports amp, and customers want to use Cohen's amp + Jamerson work together in the same dailnik. (4 speakers, 2 on Cohen and 2 on Jamerson).

Reply:
Ah I see.